### ISSUE TITLE
Support Scoped permissions to Singular Organization or Repo

### ISSUE THREAD

**Issue by aaronclong** [microsoft/vscode-copilot-release#982]
Currently Github Chat is requesting access to all repositories and organizations. This is difficult for people who use Github outside and inside of work settings. It doesn't allow them to scope Co-Pilot to their particular needs. Likewise, it creates host of legal questions on whether employers can force this product and the impact on private personal repositories/out of work activities.

https://github.com/microsoft/vscode-copilot-release/issues/813#issuecomment-1971105304

- VS Code Version: 1.87.0
- OS Version: Ventura 13.4.1 
- Logs:

Steps to Reproduce:

1. Open VsCode with latest copilot
2. Click allow on model prompt
3. See it request access to all repos and organizations

**Comment by notpeter** [comment]
Specifically, the extension is requesting 'repo' scope which includes read/access to all public, private repos and many associated settings.

![image](https://github.com/microsoft/vscode-copilot-release/assets/145113/7b2aae0a-4ced-424d-872a-a7dfd5432b24)

The reasoning provided by @TylerLeonhardt ([link](https://github.com/microsoft/vscode-copilot-release/issues/813#issuecomment-1936623781)

> We need the extra permissions (repo) in order to use a GitHub Search API because without it, it says the repo doesn't exist. GitHub doesn't split up repo permissions into read and write for this, unfortunately so we ask for the scope required to do the job.

**Comment by TylerLeonhardt** [comment]
Unfortunately, we cannot do this at this time. We use a [GitHub OAuth App](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/differences-between-github-apps-and-oauth-apps) and it doesn't and won't support the granularity of access per-repo. 

After numerous discussions with GitHub, they have recommended that we continue being a GitHub OAuth app and that GitHub Apps, which support this granularity, are not viable for us to switch to at this time.

Our hands are tied with what GitHub offers.

**Comment by aaronclong** [comment]
@TylerLeonhardt is there a ticket or issue that can tracked with Github OAuth? I understand that this might not being achieved quickly. However, this is something that should be pushed for

I am trying to phrase this in a constructive manner, but this ultimately limits the viability of CoPilot for many.  If copilot ever stops working and I can't use it without giving it total access, I fully plan to stop using CoPilot. **I know there a performance improvement doing it server-side, but can there a continued fallback to just use the local system for CoPilot?**

**Comment by sullerandras** [comment]
Could this permission be asked for at the time when the Github Copilot Chat extension needs to use the github search api? The extension seems to work fine without granting this permission.

**Comment by akshaal** [comment]
> Unfortunately, we cannot do this at this time. We use a [GitHub OAuth App](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/differences-between-github-apps-and-oauth-apps) and it doesn't and won't support the granularity of access per-repo.

You don't strictly need these permissions unless you really plan to make github copilot only to be usable within a github repostory. with no way to use it in non github repositories or in a non github workspace.

**Comment by aaronclong** [comment]
@TylerLeonhardt could a good fallback if permission is denied to use the local filesystem/plane git and/or the Github Search API? I'm interested in your thoughts here. I know the Github OAuth is lacking some **Least Privilege Access**, but I think these would be reasonable solutions to that.

Likewise, could a config be add to disable this in the vscode editor via the `settings.json`.

> Could this permission be asked for at the time when the Github Copilot Chat extension needs to use the github search api? The extension seems to work fine without granting this permission.

**Comment by aaronclong** [comment]
Any updates?

**Comment by bntzio** [comment]
> Unfortunately, we cannot do this at this time. We use a [GitHub OAuth App](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/differences-between-github-apps-and-oauth-apps) and it doesn't and won't support the granularity of access per-repo.
> 
> After numerous discussions with GitHub, they have recommended that we continue being a GitHub OAuth app and that GitHub Apps, which support this granularity, are not viable for us to switch to at this time.
> 
> Our hands are tied with what GitHub offers.
> 
> I have continued asking for a a scope that doesn't have write permission, but it's non-trivial for them so I'm working with them to come up with a good plan

If the scopes to read private repositories are necessary for the Search API with Copilot to work, and it will only be used when users with a GitHub Copilot Enterprise plan do a search via the Copilot Chat, isn't it better to ask for the extra permissions at that point?

There are two ways this can be solved:

1. Allowing users to install GitHub Copilot OR GitHub Copilot Chat (or both) and only ask for these extra permissions if they install the chat extension (and leave the basic permissions for the other extension).
2. Only require the extra permissions when triggering a Search API query with Copilot (only Enterprise users will need it, at least for now).

In my opinion, integrating those two solutions together is the way to go, but one is sufficient for now.

The easiest one to implement would be the first one, why both extensions needs to be installed?

It shouldn't be that complicated to at least give that choice to the users, while you wait for GitHub to implement more granular scopes to OAuth Apps or while users wait for improvements as the second solution I mention above or the plan you're referring to.

**Comment by missinglink** [comment]
There is literally no reason Microsoft should be asking for **write** permissions to your private work repos.

In particular the ability to modify Deploy Keys, Webhooks and Collaborators are a non-starter for any maintainer with half a brain.

I will not be using copilot until these issues are resolved and I encourage others to boycott it too.

https://en.wikipedia.org/wiki/Principle_of_least_privilege

**Comment by missinglink** [comment]
> Our hands are tied with what GitHub offers.

Doesn't MS own Github? 😆

**Comment by aaronclong** [comment]
I get Github/Microsoft wants to increase the data available to train LLMs/CoPilot. However, it really does paint copilot as insidious trojan horse. I'm wondering if developers and companies should be more concerned about copilot.

There are viable workarounds, and the lack of engagement from this teams indicates the nefarious intent!

**Comment by aaronclong** [comment]
Officially the plugin stops working unless you consent to allow access to all your repositories.
![Image](https://github.com/user-attachments/assets/c24a2770-9b26-4f67-8b29-88017e60eea9)

**Comment by wtwhite** [comment]
@TylerLeonhardt Thanks for [your comment](https://github.com/microsoft/vscode-copilot-release/issues/982#issuecomment-1971559283), in which you mention that GitHub have recommended remaining as a GitHub OAuth app instead of migrating to a GitHub App. What specifically is their/your rationale for this? Is it simply the work involved in migrating?

[The GitHub Apps best practices docs](https://docs.github.com/en/apps/creating-github-apps/about-creating-github-apps/deciding-when-to-build-a-github-app#github-apps-offer-enhanced-security) indicate that GitHub Apps can do basically everything that GitHub OAuth apps can do, except access a "GitHub enterprise object" -- is that something VS Code Copilot Chat actually needs?

If (as I suspect) it isn't, then, given the frankly crazy level of permissions being asked for by the current OAuth approach, sticking with it seems unreasonable. So if you could explain in detail why OAuth is actually necessary here, it would greatly increase our trust.

Until then, I'll be denying the request for these permissions, and encouraging others to do the same.

**Comment by indrajra** [comment]
I agree. The permissions sought is too expansive and must be restricted. 

Wonder how other IDE plugins such as "GitHub Copilot for JetBrains IDEs" for IntelliJ IDE work without requesting the permissions. Or do they too require and just did not call it out? <img width="566" alt="Image" src="https://github.com/user-attachments/assets/8d2cea97-37d1-4880-a487-b94d50563d01" />
<img width="1004" alt="Image" src="https://github.com/user-attachments/assets/77d7fcef-1276-43a7-9e80-68bab45f416f" />

**Comment by ConfusedVorlon** [comment]
This is nuts. I want to use copilot, but no way I'm granting it write access to my private repos - and those of all my clients.

**Comment by air3ijai** [comment]
Btw, can we create just a dummy GitHub account for Copilot, this is too crazy but can be a temporary workaround?
